jre/java/java/util/regex/Matcher.java

```diff
@@ -64,29 +73,43 @@ public String group() {
 
   @Override
   public String group(int group) {
-    return matchResult[group];
+    return matchResult.asArray()[group];
   }
 
   public String group(String name) {
     throw new UnsupportedOperationException();
   }
 
   public int groupCount() {
-    return null == matchResult ? 0 : matchResult.length - 1;
+    return null == matchResult ? 0 : matchResult.asArray().length - 1;
   }
 
+  // set match result, but do not advance offsets
   public boolean matches() {
+    matchResult = exec(input);
     return matchResult != null;
   }
 
   public boolean find() {
-    return null != matchResult;
+    CharSequence targetSequence = input.subSequence(offset, input.length());
+    matchResult = exec(targetSequence);
+
+    boolean isNotDone = matchResult != null;
+    if (isNotDone) {
+      prevOffset = offset;
+      offset += matchResult.getIndex() + group().length();
+    }
+    return isNotDone;
```

Comment:
It's not a good practice to have "not" in your variable name. Mostly to avoid a confusing `!isNotDone`. Instead, prefer to use the negation operator: 
```suggestion
    boolean isDone = matchResult == null;
    if (!isDone) {
      prevOffset = offset;
      offset += matchResult.getIndex() + group().length();
    }
    return isDone;
```